Added Rebecca Purple

FireCubeStudios opened the pull request:
Added Rebecca Purple color to System.Drawing.Color

Added RebeccaPurple (#663399) to KnownColorNames.cs, Color.cs, KnownColortable.cs and KnownColor.cs

#38244

FireCubeStudios committed:
Merge pull request #1 from dotnet/master

recent

FireCubeStudios committed:
Merge pull request #2 from dotnet/master

merge

FireCubeStudios committed:
Added Rebecca Purple

Added Rebecca Purple color to System.Drawing.Color

Added RebeccaPurple (#663399) to KnownColorNames.cs, Color.cs, KnownColortable.cs and KnownColor.cs

#38244

FireCubeStudios committed:
Added Rebecca Purple

Added Rebecca Purple color to System.Drawing.Color

Added RebeccaPurple (#663399) to KnownColorNames.cs, Color.cs, KnownColortable.cs and KnownColor.cs

#38244

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42785) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42785) before we can accept your contribution.<br/><br/>:x: FireCubeStudios [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42785)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=42785) it.</sub>

stephentoub commented on src/libraries/System.Drawing.Primitives/src/System/Drawing/Color.cs:
```diff
@@ -248,6 +248,8 @@ namespace System.Drawing
 
         public static Color Purple => new Color(KnownColor.Purple);
 
+        public static Color RebeccaPurple => new Color(KnownColor.RebeccaPurple);
```
Thanks.  You will need to update the ref assembly as well by adding an entry in https://github.com/dotnet/runtime/blob/9162830bc651fbff0ab1957c0155d4ecf375109e/src/libraries/System.Drawing.Primitives/ref/System.Drawing.Primitives.cs

FireCubeStudios replied:
ok i will do this soon.

stephentoub commented on src/libraries/System.Drawing.Primitives/src/System/Drawing/Color.cs:
```diff
@@ -248,6 +248,8 @@ namespace System.Drawing
         public static Color Purple => new Color(KnownColor.Purple);
 
+        public static Color RebeccaPurple => new Color(KnownColor.RebeccaPurple);
+
```
Can you also update the tests, e.g. https://github.com/dotnet/runtime/blob/9162830bc651fbff0ab1957c0155d4ecf375109e/src/libraries/System.Drawing.Primitives/tests/ColorTests.cs#L16 ?

FireCubeStudios replied:
ok i will do this soon.

FireCubeStudios committed:
Added RebeccaPurple to UnitTest and Ref

Added RebeccaPurple color to the  UnitTest and to the ref

Added RebeccaPurple a,r,g,b value to unittest in System.Drawing.Primitive. Also updated the ref.

#38244

FireCubeStudios committed:
Added RebeccaPurple to UnitTest and Ref

Added RebeccaPurple color to the  UnitTest and to the ref

Added RebeccaPurple a,r,g,b value to unittest in System.Drawing.Primitive. Also updated the ref.

#38244

safern commented on src/libraries/Common/src/System/Drawing/KnownColor.cs:
```diff
@@ -164,6 +164,7 @@ enum KnownColor
         Plum,
         PowderBlue,
         Purple,
+        RebeccaPurple,
```
This is also public, you need to add it to the ref as well:

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Drawing.Primitives/ref/System.Drawing.Primitives.cs#L193

FireCubeStudios replied:
i have added it 👍

FireCubeStudios committed:
Added RebeccaPurple to the system.drawing.primitive ref

Added RebeccaPurple color to the  System.Drawing.Primitive ref file.

"RebeccaPurple = 141" was added. I updated the other colors to their previous number by one.

#38244

FireCubeStudios committed:
Added RebeccaPurple to the system.drawing.primitive ref

Added RebeccaPurple color to the  System.Drawing.Primitive ref file.

"RebeccaPurple = 141" was added. I updated the other colors to their previous number by one.

#38244

FireCubeStudios committed:
Merge pull request #4 from FireCubeStudios/master

update

tannergooding commented on src/libraries/System.Drawing.Primitives/ref/System.Drawing.Primitives.cs:
```diff
@@ -332,40 +333,41 @@ public enum KnownColor
-        GradientInactiveCaption = 172,
-        MenuBar = 173,
-        MenuHighlight = 174,
+        RebeccaPurple = 141,
```
This is a breaking change, enum members are just constants in IL and are returned as such:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoKwgZgAJDSlHj6Bhegbyvr8YZMWAQQwYo2YAFcMMACoBDKAHMYGAM70AsgAoAlN179jhAOz1R4yTPlLVGgHTtcC9eoDcRvgF8q3oA=

So something compiled against .NET 5 using `Red` will get `RebeccaPurple` in .NET 6

tannergooding replied:
See also https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoKwgZgAJDSlHj6Bhegbyvr8YZMW7CLmgBnegFkAFAEpuvfssIB2DqIkA6AEowAJgG4lfAL5Vz1SnXowAdgFcAthrFRxVHpWV799ALz0pGgm9ABCuA4wAfTEId78AOKw9jG0FkA, for an example.

FireCubeStudios replied:
oh no, then what should i do about that? i will revert the changes. should i add rebeccapurple in the bottom with new number

tannergooding replied:
I would insert it at the bottom in the source assembly (as it needs to have the same value for `RebeccaPurple`) and then the regenerating the ref will ensure it has `= 175`

safern replied:
Good catch, I completely missed the values were swapped, I thought it was only the order.

FireCubeStudios replied:
i think this is now resolved with the latest commit i made.

tannergooding reviewed (dismissed):
`RebeccaPurple` can't be inserted as item 141, it is a breaking change.

FireCubeStudios committed:
Reverted breaking changes in drawing.primitive ref

Reverted the breaking changes in system.drawing.primivite ref.

the numbers of colors were restored back and i removed rebeccapurple.

#38244

FireCubeStudios committed:
Reverted breaking changes in drawing.primitive ref

Reverted the breaking changes in system.drawing.primivite ref.

the numbers of colors were restored back and i removed rebeccapurple.

#38244

FireCubeStudios committed:
Merge branch 'RebeccaPurple#38244' of https://github.com/FireCubeStudios/runtime into RebeccaPurple#38244

safern commented on src/libraries/Common/src/System/Drawing/KnownColor.cs:
```diff
@@ -164,6 +164,7 @@ enum KnownColor
         Plum,
         PowderBlue,
         Purple,
+        RebeccaPurple,
```
You still need to add it to the ref, but with a new int value for the enum.

tannergooding replied:
Yep and you need to ensure the value in the source matches, the easiest way to do this is to move `RebeccaPurple` to the bottom here (it can't be alphabetical unless you explicitly number all values in here which is probably undesirable)

safern replied:
Please move this to the end of the enum on this file.

FireCubeStudios replied:
done. sorry i was late was busy

FireCubeStudios committed:
Added RebeccaPurple to the ref

I added rebeccapurple to the ref file with value 175.

RebeccaPurple was added with value 175 to the bottom of the enum.

#38244

FireCubeStudios committed:
Added RebeccaPurple to the ref

I added rebeccapurple to the ref file with value 175.

RebeccaPurple was added with value 175 to the bottom of the enum.

#38244

FireCubeStudios committed:
Added rebeccapurple to bottom of Knowncolors.cs

I added rebeccapurple to theend of the file

RebeccaPurple was added to the bottom of the enum.

#38244

FireCubeStudios committed:
Added rebeccapurple to bottom of Knowncolors.cs

I added rebeccapurple to theend of the file

RebeccaPurple was added to the bottom of the enum.

#38244

FireCubeStudios commented:
> `RebeccaPurple` can't be inserted as item 141, it is a breaking change.

i think this has been fixed now

stephentoub commented on src/libraries/Common/src/System/Drawing/KnownColor.cs:
```diff
@@ -200,6 +200,7 @@ enum KnownColor
         GradientInactiveCaption,
         MenuBar,
-        MenuHighlight
+        MenuHighlight,
```
We can delete the comment above about "New system color additions...".  That's ancient history, and is no longer accurate with these additions.

FireCubeStudios replied:
i removed the comment

stephentoub reviewed (approved):
Thanks.

FireCubeStudios commented:
> `RebeccaPurple` can't be inserted as item 141, it is a breaking change.

i think this is change is fixed now

tannergooding commented:
There look to be a number of failures still, all of which appear to be caused by mismatches in the Colors tests. For example: https://dev.azure.com/dnceng/public/_build/results?buildId=864228&view=ms.vss-test-web.build-test-results-tab&runId=27680856&resultId=110085&paneView=debug

```
Assert.Equal() Failure\n                  ↓ (pos 8)\nExpected: Color [SkyBlue]\nActual:   Color [Silver]\n                  ↑ (pos 8)
```

tannergooding commented on src/libraries/Common/src/System/Drawing/KnownColorTable.cs:
```diff
@@ -124,6 +124,7 @@ internal static class KnownColorTable
             0xFFDDA0DD,     // Plum
             0xFFB0E0E6,     // PowderBlue
             0xFF800080,     // Purple
+            0xFF663399,     // RebeccaPurple
```
This is in the wrong location, it must match the order of the enum

FireCubeStudios replied:
ok

FireCubeStudios replied:
i moved it to the bottom. it should be fixed now

tannergooding commented on src/libraries/System.Drawing.Primitives/src/System/Drawing/Color.cs:
```diff
@@ -248,6 +248,8 @@ namespace System.Drawing
 
         public static Color Purple => new Color(KnownColor.Purple);
 
+        public static Color RebeccaPurple => new Color(KnownColor.RebeccaPurple);
```
nit: Would likely be beneficial to move this after MenuHighlight so it matches the enum order.

FireCubeStudios replied:
sure

FireCubeStudios replied:
there is no menuhighlight in this file. Should i move it to the end of yellow green or keep it as it is.

FireCubeStudios replied:
There is no MenuHighlight here. Should i move it to the end of yellowgreen or keep it as it is?

tannergooding replied:
I'd move it to after yellow green. @safern or I can log a bug to ensure the missing ones get added.

FireCubeStudios replied:
done

tannergooding commented on src/libraries/System.Drawing.Primitives/src/System/Drawing/KnownColorNames.cs:
```diff
@@ -149,6 +149,7 @@ internal static class KnownColorNames
             "Plum",
             "PowderBlue",
             "Purple",
+            "RebeccaPurple",
```
Same scenario, I believe this must match the enum order. This case doesn't have a corresponding comment, however and likely should above the array.

FireCubeStudios replied:
i will add the comment soon

FireCubeStudios replied:
i added a comment. I hope its good enough
"//  Names of all colors (in order of definition in the KnownColor enum)."

tannergooding commented on src/libraries/System.Drawing.Primitives/tests/ColorTests.cs:
```diff
@@ -130,6 +130,7 @@ public class ColorTests
                 new object[] {"Plum", 255, 221, 160, 221},
                 new object[] {"PowderBlue", 255, 176, 224, 230},
                 new object[] {"Purple", 255, 128, 0, 128},
+                new object[] {"RebeccaPurple", 255, 102, 51, 153},
```
You should check if the tests have the same requirement as well.

FireCubeStudios replied:
will check to see if unit tests will have the requirement soon

FireCubeStudios replied:
i was unable to check tests today because of visual studio error :/
but i will do it hopefully tomorow.

FireCubeStudios committed:
Fixed some of the requested changes with ordering of Enum

Removed a comment in "knowncolor.cs" and moved rebeccapurple to match enum order.

moved rebeccapurple in knowncolortable.cs, color.cs, and knowncolornames.cs

#38244

FireCubeStudios committed:
Merge branch 'RebeccaPurple#38244' of https://github.com/FireCubeStudios/runtime into RebeccaPurple#38244

safern commented on src/libraries/Common/src/System/Drawing/KnownColorTable.cs:
```diff
@@ -151,6 +152,7 @@ internal static class KnownColorTable
             0xFFF5F5F5,     // WhiteSmoke
             0xFFFFFF00,     // Yellow
             0xFF9ACD32,     // YellowGreen
+            0xFF663399,     // RebeccaPurple
```
You forgot to remove the entry above that you added before, so this is now duplicate.

safern commented on src/libraries/System.Drawing.Primitives/src/System/Drawing/Color.cs:
```diff
@@ -247,6 +247,10 @@ namespace System.Drawing
+        
+        public static Color RebeccaPurple => new Color(KnownColor.RebeccaPurple);
+
+        public static Color RebeccaPurple => new Color(KnownColor.RebeccaPurple);
```
This is duplicated, missed to remove the previous one.

FireCubeStudios replied:
fixed thanks

safern commented on src/libraries/System.Drawing.Primitives/src/System/Drawing/KnownColorNames.cs:
```diff
@@ -149,6 +149,7 @@ internal static class KnownColorNames
             "Plum",
             "PowderBlue",
             "Purple",
+            "RebeccaPurple",
```
Remove this line, this is duplicate, seems like forgot to remove it when moved to the end of the list.

FireCubeStudios replied:
fixed thanks

FireCubeStudios committed:
Removed duplicates and added a comment

Added a comment in KnownColorNames.cs stating that the array follows order of knowncolor enum. Also removed duplicate rebeccapurple entries

#38244

tannergooding commented on src/libraries/Common/src/System/Drawing/KnownColorTable.cs:
```diff
@@ -151,6 +151,7 @@ internal static class KnownColorTable
             0xFFF5F5F5,     // WhiteSmoke
             0xFFFFFF00,     // Yellow
             0xFF9ACD32,     // YellowGreen
+            0xFF663399,     // RebeccaPurple
```
I've looked at the current test failures and roughly root caused them.

Basically, before this change we had a large list of colors that were added in (at least as far as the docs tell me) full framework v1.1. These include some System Colors from 1 to 26, inclusive, and "Web" Colors from 27 to 167. Full framework v2 then added a few additional system colors into the mix starting at 168.

`s_colorTable` only contains the "Web" colors (starting from `Transparent` at 27). `ArgbToKnownColor` iterates this list in its entirety and will end up mismatching for `RebeccaPurple` because it will end up returning `ButtonFace` for `RebeccaPurple`. It needs to be updated to account for this gap.

The same is true for `KnownColorToArgb` and `IsKnownColorSystem`. `GetSystemColorArgb` should likely add in an assert that it is handled correctly as well.

You are of course free to push these fixes yourself, but I can also push any necessary changes/commits if you'd prefer that instead.

CC. @safern.

FireCubeStudios replied:
ok i will try to attempt to solve it. visual studio has been really weird lately so i couldnt check earlier

FireCubeStudios replied:
update I unfortunately couldnt manage to get it working. If you are able to can you please push the changes or maybe someone else

safern replied:
@tannergooding since you already have a hold of what's going on, could you help @FireCubeStudios ?

tannergooding replied:
Yep, will do.

tannergooding replied:
Should be resolved now. All tests passed locally.

safern replied:
Thanks, @tannergooding

tannergooding committed:
Fixing up the handling of RebeccaPurple

safern commented on src/libraries/Common/src/System/Drawing/KnownColorTable.cs:
```diff
@@ -163,7 +169,14 @@ internal static Color ArgbToKnownColor(uint argb)
-                    return Color.FromKnownColor((KnownColor)(index + (int)KnownColor.Transparent));
+                    var knownColor = KnownColor.Transparent + index;
+
+                    if (knownColor > KnownColor.YellowGreen)
```
Maybe add a comment on why this is needed?

FireCubeStudios replied:
i added a comment to it. should i also add a comment to the other ones? like on these codes:
```suggestion
                   return color < KnownColor.ButtonFace
                 ? s_colorTable[(int)color - (int)KnownColor.Transparent]
                 : s_colorTable[(int)color - (int)KnownColor.RebeccaPurple + ((int)KnownColor.YellowGreen - (int)KnownColor.WindowText)];
```

safern replied:
I think it is good with the one you added, thanks.

FireCubeStudios committed:
Added a comment to explain changes

Added a comment in KnownColortable.cs stating that the code accounts for the system colors and new rebeccapurple color
#38244

safern commented:
/azp run runtime
